security/tasks.py

```diff
@@ -0,0 +1,113 @@
+import os
+import base64
+import logging
+import pickle
+import sys
+
+from django.conf import settings
+from django.core.management import call_command
+from django.core.exceptions import ImproperlyConfigured
+from django.db import transaction
+from django.utils.timezone import now
+
+try:
+    from celery import Task
+    from celery import shared_task
+except ImportError:
+    raise ImproperlyConfigured('Missing celery library, please install it')
+
+from chamber.shortcuts import change_and_save
+from chamber.utils.transaction import atomic_with_signals
+
+from .models import CeleryTaskLog, CeleryTaskLogState
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+class LoggedTask(Task):
+
+    abstract = True
+    logger_level = logging.WARNING
+
+    def get_task(self, task_id):
+        return CeleryTaskLog.objects.get(pk=task_id)
+
+    def __call__(self, *args, **kwargs):
+        # Every set attr is send here
+        self.on_start(self.request.id, args, kwargs)
+        super().__call__(*args, **kwargs)
+
+    def _call_callback(self, event, *args, **kwargs):
+        if hasattr(self, 'on_{}_callback'.format(event)):
+            getattr(self, 'on_{}_callback'.format(event))(*args, **kwargs)
+
+    def on_apply(self, task_id, args, kwargs):
+        self._call_callback('apply', task_id, args, kwargs)
+
+    def on_start(self, task_id, args, kwargs):
+        self.get_task(task_id).change_and_save(state=CeleryTaskLogState.ACTIVE, start=now())
+        self._call_callback('start', task_id, args, kwargs)
+
+    def on_success(self, retval, task_id, args, kwargs):
+        self.get_task(task_id).change_and_save(state=CeleryTaskLogState.SUCCEEDED, stop=now())
+        self._call_callback('success', task_id, args, kwargs)
+
+    def on_failure(self, exc, task_id, args, kwargs, einfo):
+        try:
+            self.get_task(task_id).change_and_save(state=CeleryTaskLogState.FAILED, stop=now(), error_message=einfo)
+        except CeleryTask.DoesNotExist:
+            pass
+        self._call_callback('failure', task_id, args, kwargs)
+
+    def apply_async_on_commit(self, args=None, kwargs=None, **options):
+        task = CeleryTaskLog.objects.create(name=self.name, state=CeleryTaskLogState.WAITING,
+                                            queue_name=options.get('queue', settings.CELERY_DEFAULT_QUEUE))
+        args = (task.pk,) + tuple(args or ())
+        self.on_apply(task.pk, args, kwargs)
+        if sys.argv[1:2] == ['test']:
+            super().apply_async(args=args, kwargs=kwargs, task_id=task.pk, **options)
```

Comment:
everywhere else you are casting the task.pk to str, shouldn't you do the same here?

Reply:
yes thanks